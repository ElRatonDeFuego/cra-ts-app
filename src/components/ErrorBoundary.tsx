import { faTriangleExclamation } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { Component as ReactComponent } from "react";

interface ErrorBoundaryProps {
  children: React.ReactNode;
  fallback?: React.ReactNode;
}

interface ErrorBoundaryState {
  hasError: boolean;
}

export class ErrorBoundary extends ReactComponent<
  ErrorBoundaryProps,
  ErrorBoundaryState
> {
  /* eslint-disable no-restricted-syntax */
  public constructor(props: ErrorBoundaryProps) {
    super(props);
    this.state = { hasError: false };
  }

  public static getDerivedStateFromError = () => ({ hasError: true });

  // eslint-disable-next-line class-methods-use-this
  public override componentDidCatch = (error: unknown, errorInfo: unknown) => {
    // eslint-disable-next-line no-console
    console.error("ErrorBoundary caught this:", error, errorInfo);
  };

  // eslint-disable-next-line @typescript-eslint/member-ordering
  public override render() {
    if (!this.state.hasError) {
      return <>{this.props.children}</>;
    }

    return this.props.fallback ? (
      this.props.fallback
    ) : (
      <div className="min-h-screen flex items-center justify-center">
        <div
          className="flex content-start items-center
            animate-[wiggle_0.15s_ease-in] border-t-2 border-b-2 border-red-200"
        >
          <FontAwesomeIcon
            className="text-6xl text-red-400 p-10"
            icon={faTriangleExclamation}
          />
          <div className="text-gray-800 p-20 leading-loose text-2xl">
            We are unable to display this page at the moment.
            <br />
            Please try again later.
          </div>
        </div>
      </div>
    );
  }
  /* eslint-enable no-restricted-syntax */
}
