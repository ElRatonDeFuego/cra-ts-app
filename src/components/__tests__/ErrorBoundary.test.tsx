import { useLayoutEffect } from "react";
import renderer from "react-test-renderer";
import { ErrorBoundary } from "../ErrorBoundary";

const ComponentThatWillThrow = () => {
  useLayoutEffect(() => {
    throw new Error("will be caught by the error boundary");
  });

  return <>this will never render</>;
};

test("should match the snapshot without a fallback", () => {
  const tree = renderer
    .create(
      <ErrorBoundary>
        <ComponentThatWillThrow />
      </ErrorBoundary>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test("should match the snapshot with a fallback", () => {
  const tree = renderer
    .create(
      <ErrorBoundary fallback={<>the fallback component</>}>
        <ComponentThatWillThrow />
      </ErrorBoundary>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
