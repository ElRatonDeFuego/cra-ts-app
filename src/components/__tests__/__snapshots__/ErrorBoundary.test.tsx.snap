// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot with a fallback 1`] = `"the fallback component"`;

exports[`should match the snapshot without a fallback 1`] = `
<div
  className="min-h-screen flex items-center justify-center"
>
  <div
    className="flex content-start items-center animate-[wiggle_0.15s_ease-in] border-t-2 border-b-2 border-red-200"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-triangle-exclamation text-6xl text-red-400 p-10"
      data-icon="triangle-exclamation"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M256 32c14.2 0 27.3 7.5 34.5 19.8l216 368c7.3 12.4 7.3 27.7 .2 40.1S486.3 480 472 480H40c-14.3 0-27.6-7.7-34.7-20.1s-7-27.8 .2-40.1l216-368C228.7 39.5 241.8 32 256 32zm0 128c-13.3 0-24 10.7-24 24V296c0 13.3 10.7 24 24 24s24-10.7 24-24V184c0-13.3-10.7-24-24-24zm32 224c0-17.7-14.3-32-32-32s-32 14.3-32 32s14.3 32 32 32s32-14.3 32-32z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
    <div
      className="text-gray-800 p-20 leading-loose text-2xl"
    >
      We are unable to display this page at the moment.
      <br />
      Please try again later.
    </div>
  </div>
</div>
`;
